draco/core/io.py

```diff
@@ -352,6 +342,328 @@ def process(self):
+            # (eq. 7 in du Mas des Bourboux et al)
+
+            # can't just concatenate because lam axis is different
+            num_objects = len(pos)
+            num_pix = len(lam)
+
+            self.log.debug("rank 0 done reading files")
+            self.log.debug(f"delta shape {delta.shape}")
+
+        else:
+            num_objects = None
+            num_pix = None
+            delta = None
+            weight = None
+
+        self.comm.Barrier()
+
+        # Broadcast the size of the catalog to all ranks
+        num_objects = self.comm.bcast(num_objects, root=0)
+        num_pix = self.comm.bcast(num_pix, root=0)
+        self.log.debug(f"Reading spectra for {num_objects} objects.")
+
+        # broadcast the axes
+        if self.comm.rank != 0:
+            pos = np.zeros((num_objects, 2), dtype=np.float64)
+            z = np.zeros(num_objects, dtype=np.float64)
+            lam = np.zeros(num_pix, dtype=np.float64)
+            obj_id = np.zeros(num_objects, dtype=np.int32)
+        self.comm.Bcast(pos, root=0)
+        self.comm.Bcast(z, root=0)
+        self.comm.Bcast(lam, root=0)
+        self.comm.Bcast(obj_id, root=0)
+        self.log.debug(f"(rank {self.comm.rank}) got axes.")
+
+        # convert wavelength (A) to frequency (MHz)
+        freq = units.c / lam * 1e4
+
+        # scatter the distributed datasets over frequency
+        def scatter_array(array, root=0, axis=0):
+            # split distributed axis
+            if self.comm.rank == root:
+                shape = array.shape
+                n = shape[axis]
+                dtype = array.dtype
+            else:
+                shape = None
+                n = None
+                dtype = None
+            shape = self.comm.bcast(shape, root=root)
+            dtype = self.comm.bcast(dtype, root=root)
+            mpitype = mpiutil.typemap(dtype)
+            n = self.comm.bcast(n, root=root)
+            n_all, s_all, e_all = mpiutil.split_all(n, comm=self.comm)
+            slices = []
+            for r in range(self.comm.size):
+                this_slice = [slice(None)] * len(shape)
+                this_slice[axis] = slice(s_all[r], e_all[r])
+                slices.append(tuple(this_slice))
+
+            # scatter array from source rank
+            if self.comm.rank == root:
+                local = array[slices[root]]
+                # send to each rank
+                sends = []
+                for r in range(self.comm.size):
+                    if r == root:
+                        continue
+                    to_send = array[slices[r]].flatten()
+                    sends.append(self.comm.Isend([to_send, mpitype], dest=r))
+                for req in sends:
+                    req.wait()
+            else:
+                local_shape = list(shape)
+                local_shape[axis] = n_all[self.comm.rank]
+                local = np.zeros(local_shape, dtype=dtype)
+                self.comm.Recv([local, mpitype], source=root)
```

Comment:
You typically don't need to give the mpitype in these send/recv calls. They'll get inferred.